### ISSUE TITLE
Add hwaccel APIs

### ISSUE THREAD

**Issue by mariodivece** [Ruslan-B/FFmpeg.AutoGen#59]
@Ruslan-B Could you please add the hwaccel.h reference to the project? 
User @iwarp has done it and provided me with his project files: 
https://github.com/unosquare/ffmediaelement/files/1263051/FFmpeg.AutoGen.zip

It would be great if you could merge these changes so hwaccel APIs become available for everyone.
For additional background info please refer to:
https://github.com/unosquare/ffmediaelement/issues/69

**Comment by Ruslan-B** [comment]
@mariodivece 
It looks like it is relatively easy to add this API.
However, I have few question maybe you or @iwarp can help me with depends from answers it might speed up the process :
1.  I have this in Windows Kits:
C:\Program Files (x86)\Windows Kits\10\Include\10.0.15063.0\ucrt
But I don't have these:
C:\Program Files (x86)\Windows Kits\10\Include\10.0.15063.0\shared
C:\Program Files (x86)\Windows Kits\10\Include\10.0.15063.0\um
Do you know which part of Windows Kits they are belongs to?

2. Do you know origin of mfx include files?

**Comment by iwarp** [comment]
Hi @Ruslan-B,
I had enabled Intel QSV, Cuda Video, DXVA and D3D11 in order to directly access the GPU textures returned from the decode. I'm currently trying to get the texture to draw directly via D3DImage without the CPU based pixel format changes to further reduce CPU playback. Some of these can be omitted as we and leave the FFmpeg avutil to abstract that layer away.

1. This is for the DXVA and D3D11 headers files. I think these come from the UWP Kit which i had enabled in Visual Studio for another project. If that doesn't work let me know and i'll keep digging.
C:\Program Files (x86)\Windows Kits\10\Include\10.0.15063.0\shared
C:\Program Files (x86)\Windows Kits\10\Include\10.0.15063.0\um


2. MFX file is part of the Intel Media SDK for QSV https://software.intel.com/en-us/media-sdk, not 100% sure of the license with this.

3. I also update one or two of the header files in the FFmpeg Include mainly for DX11 support.

**Comment by Ruslan-B** [comment]
Hi @iwarp, thanks for update please keep me posted wrt further progress. Meanwhile, I'll try to figure out the best way of integration.

**Comment by iwarp** [comment]
Hi @Ruslan-B,
I've managed to get D3D11 decoding and displaying the Texture2D. I don't think there is anything else i need to add. 

What needs to be done to get this added?

**Comment by mariodivece** [comment]
Yep. Also waiting on this to include it in FFME. Thanks @Ruslan-B and @iwarp

**Comment by Ruslan-B** [comment]
@iwarp @mariodivece sorry folks hadn't much time recently. I'm looking right now and trying to workaround need of ffmpeg headers modification. @iwarp any hits or at-least origin of these changes?

**Comment by Ruslan-B** [comment]
Folks, I added D3D11 hwcontext support without modifying any headers. Is it enough? Or you still would like to mfx support?

**Comment by mariodivece** [comment]
Is dxva2 (d3d 9) included with this change? I believe dxva2 is the most common. My plan is to try to create a d3d11 hw context and if this fails then create a d3d 9 context. If all hw fails, fallback to software only.

**Comment by Ruslan-B** [comment]
yes, it is included. I added libavutil/hwcontext.h, libavutil/hwcontext_dxva2.h, libavcodec/dxva2.h, libavcodec/d3d11va.h.
If something missing just let me know.

**Comment by mariodivece** [comment]
@Ruslan-B AWESOME!! 
@iwarp could you provide me with your latest usage code so I can integrate into FFME?

**Comment by iwarp** [comment]
@Ruslan-B  No i think we can drop MFX, i think i'm only going to be using DXVA2 and D3D11. So long as pixfmt.h has also been updated to use the new D3D11 pixel format changes. Now i understand how the hwaccel works getting access to the individual frames.

**Comment by Ruslan-B** [comment]
nice, closing this issue then.

**Comment by Ruslan-B** [comment]
I created nuget package as well. It is without any ffmpeg binaries.
Install-Package FFmpeg.AutoGen -Version 3.3.3.1

**Comment by iwarp** [comment]
@Ruslan-B I've just tried your nuget package, it seems it's missing two D3D11 features everything seems to be working in my project. @mariodivece FYI

AVPixelFormat.AV_PIX_FMT_D3D11
AVHWDeviceType.AV_HWDEVICE_TYPE_D3D11VA

There are both are missing from your repo they need to be refreshed from the FFmpeg project.
avutil\pixfmt.h
avutil\hwcontext.h

**Comment by mariodivece** [comment]
correct. I added a separate issue #61

**Comment by mariodivece** [comment]
@iwarp could you please look at my issue https://github.com/unosquare/ffmediaelement/issues/69
I tried to get DXVA2 to work and it's not working. :(

**Comment by Xhone** [comment]
Hi @Ruslan-B 
 I would like to know how  the hwaccel works with DXVA2 using FFmpeg.AutoGen
